command/agent/cache/api_proxy.go

```diff
@@ -0,0 +1,58 @@
+package cache
+
+import (
+	"bytes"
+	"context"
+	"io/ioutil"
+
+	hclog "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/vault/api"
+)
+
+// APIProxy is an implementation of the proxier interface that is used to
+// forward the request to Vault and get the response.
+type APIProxy struct {
+	logger hclog.Logger
+}
+
+type APIProxyConfig struct {
+	Logger hclog.Logger
+}
+
+func NewAPIProxy(config *APIProxyConfig) Proxier {
+	return &APIProxy{
+		logger: config.Logger,
+	}
+}
+
+func (ap *APIProxy) Send(ctx context.Context, req *SendRequest) (*SendResponse, error) {
+	client, err := api.NewClient(api.DefaultConfig())
+	if err != nil {
+		return nil, err
+	}
+	client.SetToken(req.Token)
+	client.SetHeaders(req.Request.Header)
+
+	fwReq := client.NewRequest(req.Request.Method, req.Request.URL.Path)
+	fwReq.BodyBytes = req.RequestBody
+
+	// Make the request to Vault and get the response
+	ap.logger.Info("forwarding request", "path", req.Request.URL.Path)
+	resp, err := client.RawRequestWithContext(ctx, fwReq)
+	if err != nil {
+		return nil, err
+	}
+
+	// Parse and reset response body
+	respBody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		ap.logger.Error("failed to read request body", "error", err)
+		return nil, err
+	}
```

Comment:
We shouldn't be closing the response body here as it will be read out up in the http layer. I saw that it wasn't getting closed there as well. Added that as a `defer` there. Kindly take a look at the recent commit that does that.

Reply:
But we are creating new read closers, so we are closing the new ones but not the existing one?

Reply:
Possibly. I've fixed this in c4c930ad52dc6e238d6dea3fa8bbce3b6dd1833f. Kindly take a look.